<a href="https://colab.research.google.com/github/Tristan-Brown1096/DS-Unit-1-Sprint-2-Statistics/blob/master/module2/LS_DS_122_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [1]:
# YOUR WORK HERE
#importing important libraries
import pandas as pd
import numpy as np
from scipy import stats

#obtaining dataset
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
!unzip student.zip

#creating dataframe
df = pd.read_csv('student-mat.csv', sep=';')
df.head()

--2020-07-14 18:45:43--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.1s    

2020-07-14 18:45:44 (145 KB/s) - ‘student.zip’ saved [20478/20478]

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [12]:
# YOUR WORK HERE
#creating contingency tables
contingency_1 = pd.crosstab(df['romantic'], df['nursery'])
contingency_2 = pd.crosstab(df['traveltime'], df['reason'])
contingency_3 = pd.crosstab(df['freetime'], df['goout'])
contingency_4 = pd.crosstab(df['address'], df['internet'])

## $X^2$ Test 1

1) Null Hypothesis:

Having a romantic relationship and attending nursery school are independent of each other

2) Alternative Hypothesis:

Having a romantic relationship and attending nursery school have a dependent relationship

3) Confidence Level: 95%

In [13]:
#printing 1st contingency table
contingency_1

nursery   no  yes
romantic         
no        56  207
yes       25  107

In [14]:
#performing 1st chi^2 test
chi2_1, p1, dof1, ex1 = stats.chi2_contingency(contingency_1)
print("chi2 statistic:", chi2_1)
print("p value:", p1)
print("degrees of freedom:",dof1)
print("expected frequencies table:\n", ex1)

chi2 statistic: 0.1716861929173629
p value: 0.6786175620713879
degrees of freedom: 1
expected frequencies table:
 [[ 53.93164557 209.06835443]
 [ 27.06835443 104.93164557]]


4) Based on a $X^2$ statistic of 0.17 and a p-value of 0.68, we fail to reject the null hypothesis that the variables are independent

## $X^2$ Test 2

1) Null Hypothesis:

Travel time and reason for attending the school are independent of each other

2) Alternative Hypothesis:

Travel time and reason for attending the school have a dependent relationship

3) Confidence Level: 95%

In [15]:
#printing 2nd contingency table
contingency_2

reason      course  home  other  reputation
traveltime                                 
1               90    77     23          67
2               37    27     10          33
3               12     3      3           5
4                6     2      0           0

In [16]:
#performing 2nd chi^2 test
chi2_2, p2, dof2, ex2 = stats.chi2_contingency(contingency_2)
print("chi2 statistic:", chi2_2)
print("p value:", p2)
print("degrees of freedom:",dof2)
print("expected frequencies table:\n", ex2)

chi2 statistic: 11.765560860595862
p value: 0.22685085821004924
degrees of freedom: 9
expected frequencies table:
 [[94.34177215 70.91898734 23.42278481 68.3164557 ]
 [39.27848101 29.52658228  9.75189873 28.44303797]
 [ 8.44303797  6.34683544  2.09620253  6.11392405]
 [ 2.93670886  2.20759494  0.72911392  2.12658228]]


Based on a $X^2$ statistic of 11.77 and a p-value of 0.23, we fail to reject the null hypothesis that the variables are independent.

## $X^2$ Test 3

1) Null Hypothesis:

The amount of free time a student has and the frequency at which they go out with friends are independent of each other

2) Alternative Hypothesis:

The amount of free time a student has and the frequency at which they go out with friends have a dependent relationship

3) Confidence Level: 95%

In [17]:
#printing 3rd contingency table
contingency_3

goout     1   2   3   4   5
freetime                   
1         4   8   5   1   1
2         5  27  17  11   4
3         5  41  70  26  15
4         5  18  34  41  17
5         4   9   4   7  16

In [10]:
#performing 3rd chi^2 test
chi2_3, p3, dof3, ex3 = stats.chi2_contingency(contingency_3)
print("chi2 statistic:", chi2_3)
print("p value:", p3)
print("degrees of freedom:",dof3)
print("expected frequencies table:\n", ex3)

chi2 statistic: 80.87785174188721
p value: 1.155517403684047e-10
degrees of freedom: 16
expected frequencies table:
 [[ 1.10632911  4.95443038  6.25316456  4.13670886  2.54936709]
 [ 3.72658228 16.68860759 21.06329114 13.93417722  8.58734177]
 [ 9.14177215 40.93924051 51.67088608 34.18227848 21.06582278]
 [ 6.69620253 29.98734177 37.84810127 25.03797468 15.43037975]
 [ 2.32911392 10.43037975 13.16455696  8.70886076  5.36708861]]


Based on a $X^2$ statistic of 80.88 and a p-value of 0.0000000001, we reject the null hypothesis that the amount of free time a student has and the frequency at which they go out with friends are independent.

## $X^2$ Test 4

1) Null Hypothesis:

Internet access and home address type are independent of each other

2) Alternative Hypothesis:

Internet access and home address type have a dependent relationship

3) Confidence Level: 95%

In [18]:
#printing 4th contingency table
contingency_4

internet  no  yes
address          
R         28   60
U         38  269

In [29]:
#performing 4th chi^2 test
chi2_4, p4, dof4, ex4 = stats.chi2_contingency(contingency_4)
print("chi2 statistic:", chi2_4)
print("p value:", p4)
print("degrees of freedom:",dof4)
print("expected frequencies table:\n", ex4)

chi2 statistic: 17.202541580860636
p value: 3.3598654758469596e-05
degrees of freedom: 1
expected frequencies table:
 [[ 14.70379747  73.29620253]
 [ 51.29620253 255.70379747]]


Based on a $X^2$ statistic of 17.20 and a p-value of 0.00003, we can reject the null hypothesis that internet access and home address type are independent.

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [30]:
# YOUR WORK HERE
pd.crosstab(df['address'], df['internet'], margins=True)

internet  no  yes  All
address               
R         28   60   88
U         38  269  307
All       66  329  395

In [31]:
observed = contingency_4
observed

internet  no  yes
address          
R         28   60
U         38  269

In [37]:
row_totals = df['internet'].value_counts().sort_index().values
row_totals

array([ 66, 329])

In [38]:
col_totals = df['address'].value_counts().sort_index().values
col_totals

array([ 88, 307])

In [56]:
sample_size = df.shape[0]
sample_size

395

In [44]:
expected = np.array([[0.0,0.0],
                     [0.0,0.0]])

for row_position, col in enumerate(col_totals):
  for column_position, row in enumerate(row_totals):
    expected_value = (row*col) / sample_size
    expected[row_position][column_position] = expected_value

expected

array([[ 14.70379747,  73.29620253],
       [ 51.29620253, 255.70379747]])

In [62]:
chi2 = (((abs(observed - expected) - 0.5) ** 2) / expected).sum().sum()
chi2

17.202541580860636

In [41]:
chi2_4, p4, dof4, ex4 = stats.chi2_contingency(contingency_4)
print("chi2 statistic:", chi2_4)
print("p value:", p4)
print("degrees of freedom:",dof4)
print("expected frequencies table:\n", ex4)

chi2 statistic: 17.202541580860636
p value: 3.3598654758469596e-05
degrees of freedom: 1
expected frequencies table:
 [[ 14.70379747  73.29620253]
 [ 51.29620253 255.70379747]]


## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [116]:
# YOUR WORK HERE
def chi_2_calc(a1, a2, d):
  obs = pd.crosstab(a1, a2)
  rt = a2.value_counts().sort_index().values
  ct = a1.value_counts().sort_index().values
  ss = d.shape[0]
  exp = np.ndarray([a1.value_counts().shape[0], a2.value_counts().shape[0]], dtype=float)
  for rp, col in enumerate(ct):
    for cp, row in enumerate(rt):
      ev = (row*col) / sample_size
      exp[rp][cp] = ev
  if exp.shape == (2,2):
    return (((abs(obs - exp) - 0.5) ** 2) / exp).sum().sum()
  else:
    return ((((obs - exp)) ** 2) / exp).sum().sum()

In [117]:
chi_2_calc(df['freetime'], df['goout'], df)

80.8778517418872

In [114]:
stats.chi2_contingency(contingency_3)

(80.87785174188721,
 1.155517403684047e-10,
 16,
 array([[ 1.10632911,  4.95443038,  6.25316456,  4.13670886,  2.54936709],
        [ 3.72658228, 16.68860759, 21.06329114, 13.93417722,  8.58734177],
        [ 9.14177215, 40.93924051, 51.67088608, 34.18227848, 21.06582278],
        [ 6.69620253, 29.98734177, 37.84810127, 25.03797468, 15.43037975],
        [ 2.32911392, 10.43037975, 13.16455696,  8.70886076,  5.36708861]]))



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

A sample "estimate" is the sample mean, which is used to attempt to estimate the population mean.

- What are hypothesis test? How are they useful?

A hypothesis test is used to test various hypotheses to attempt to reject the null hypothesis and have confidence in the alternative.

- What is a "null hypothesis?"

The null hypothesis is the default assumption of the experiment.

- What is a p-value and what does it represent?

The p-value is the likelihood that the null hypothesis is non-rejectable.

- What does it mean for something to be "statistically significant?"

A finding is statistically significant when the likelihood of the alternative hypothesis is true is 95% or greater, represented by a p-value of 0.05.

- What is a test statistic and how does it relate to a p-value?

A test statistic is the number resulting from a hypothesis test. These values correspond to a specific p-value based on a table depending on the particular hypothesis test.

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

Degrees of freedom refers to the number of values that can vary freely in a set of information. For a 1 or 2 sample t-test, the degrees of freedom is equal to the sample size-1. For a $X^2$ test for independence, the degrees of freedom is equal to the number of rows-1 times the number of columns-1.

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)